File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head)
    {
        stack<int> st;
        ListNode* temp = head;
        while(temp != NULL)
        {
            st.push(temp->val);
            temp=temp->next;
        }

        int n = st.size();
        int i=0;
        int sum=INT_MIN;
        temp = head;
        while(temp != NULL && i != n/2)
        {
            sum = max(sum,(temp->val+st.top()));
            i++;
            temp=temp->next;
            st.pop();
        }
        return sum;
    }
};

File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.java
class Solution {
    public int pairSum(ListNode head)
    {
        ListNode slow = head;
        ListNode fast = head;
        ListNode prev = null;

        // Reverse first half
        while(fast != null && fast.next != null)
        {
            fast = fast.next.next;

            ListNode next = slow.next;
            slow.next = prev;
            prev = slow;
            slow = next;
        }

        // prev = reversed first half
        // slow = second half
        int ans = 0;

        while(prev != null && slow != null)
        {
            ans = Math.max(ans, prev.val + slow.val);

            prev = prev.next;
            slow = slow.next;
        }

        return ans;
    }
}